hls-plugin-api/src/Ide/Plugin/Properties.hs

```diff
@@ -0,0 +1,341 @@
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE ConstraintKinds      #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE GADTs                #-}
+{-# LANGUAGE KindSignatures       #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE RecordWildCards      #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+-- See Note [Constraints]
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Ide.Plugin.Properties
+  ( PropertyType (..),
+    ToHsType,
+    MetaData (..),
+    PropertyKey,
+    SPropertyKey (..),
+    Properties,
+    HasProperty,
+    emptyProperties,
+    defineNumberProperty,
+    defineStringProperty,
+    defineBooleanProperty,
+    defineObjectProperty,
+    defineArrayProperty,
+    defineEnumProperty,
+    toDefaultJSON,
+    toVSCodeExtensionSchema,
+    usePropertyEither,
+    useProperty,
+    (&),
+  )
+where
+
+import qualified Data.Aeson       as A
+import qualified Data.Aeson.Types as A
+import           Data.Either      (fromRight)
+import           Data.Function    ((&))
+import           Data.Kind        (Constraint)
+import qualified Data.Map.Strict  as Map
+import           Data.Proxy       (Proxy (..))
+import qualified Data.Text        as T
+import           GHC.TypeLits
+import           Unsafe.Coerce    (unsafeCoerce)
+
+-- ---------------------------------------------------------------------
+
+data PropertyType
+  = TNumber
+  | TString
+  | TBoolean
+  | TObject
+  | TArray
+  | TEnum
+
+type family ToHsType (t :: PropertyType) where
+  ToHsType 'TNumber = Int
+  ToHsType 'TString = T.Text
+  ToHsType 'TBoolean = Bool
+  ToHsType 'TObject = A.Object
+  ToHsType 'TArray = A.Array
+  ToHsType 'TEnum = T.Text -- supports only text enum now
+
+-- ---------------------------------------------------------------------
+
+data MetaData (t :: PropertyType) where
+  MetaData ::
+    (IsNotTEnum t) =>
+    {defaultValue :: ToHsType t, description :: T.Text} ->
+    MetaData t
+  EnumMetaData ::
+    (IsTEnum t) =>
+    { defaultValue :: ToHsType t,
+      description :: T.Text,
+      enumValues :: [T.Text],
+      enumDescriptions :: [T.Text]
+    } ->
+    MetaData t
+
+data PropertyKey (s :: Symbol) (t :: PropertyType)
```

Comment:
Promotion would make this cleaner:

```suggestion
data PropertyKey = PropertyKey Symbol PropertyType
```

and now you can give your rows kind `[PropertyKey]` instead of `[Type]`

Reply:
Sorry, I don't understand clearly. How does it work?

Reply:
The change above means you can change the `r` parameter from `r :: [*]` to `r :: [PropertyKey]`, which much better documents how your row type works. Now you can use the promoted `'PropertyKey` constructor exactly how you were using the `data PropertyKey` in the old code (I don't think it will take any changes.)

Reply:
Wow, that's cool! Many thanks, today I learnd data consturctor with parameters can also be promoted